hal/src/common/gpio/v2/pin.rs

```diff
@@ -842,7 +1032,49 @@ where
         }
     }
 
-    /// Convert the pin to the requested [PinMode]
+    /// Convert a [`Pin`] to a type that implements [`AnyPin`]
+    ///
+    /// Even though there is a one-to-one mapping between `Pin<I, M>` and
+    /// `AnyPin<Id = I, Mode = M>`, the compiler doesn't know that. This method
+    /// provides a way to convert from a [`Pin`] to an [`AnyPin`]. See the
+    /// [`AnyPin`] trait for more details.
+    #[inline]
+    pub fn as_any<P>(self) -> P
+    where
+        P: AnyPin<Id = I, Mode = M>,
+    {
+        // SAFETY:
+        // core::ptr::read makes a bitwise copy, regardless of whether the type
+        // implements Copy. Thus, the returned value is a copy, and we still
+        // need to drop self.
+        unsafe { core::ptr::read(&self as *const _ as *const P) }
```

Comment:
Could someone please review this `unsafe` code and make sure you agree with my safety justification. For it to make sense, you'll probably need to understand the example in the `AnyPin` documentation above.